src/librustc_mir/transform/simplify_try.rs

```diff
@@ -0,0 +1,201 @@
+//! The general point of the optimizations provided here is to simplify something like:
+//!
+//! ```rust
+//! match x {
+//!     Ok(x) => Ok(x),
+//!     Err(x) => Err(x)
+//! }
+//! ```
+//!
+//! into just `x`.
+
+use crate::transform::{MirPass, MirSource, simplify};
+use rustc::ty::{TyCtxt, Ty};
+use rustc::mir::*;
+use rustc_target::abi::VariantIdx;
+use itertools::Itertools as _;
+
+/// Simplifies arms of form `Variant(x) => Variant(x)` to just a move.
+///
+/// This is done by transforming basic blocks where the statements match:
+///
+/// ```rust
+/// _LOCAL_TMP = ((_LOCAL_1 as Variant ).FIELD: TY );
+/// ((_LOCAL_0 as Variant).FIELD: TY) = move _LOCAL_TMP;
```

Comment:
This is only sound to collapse if `_LOCAL_TMP` isn't used *anywhere* else (see `promote_consts` for an example of an analysis like that).

But also, there's another issue here... namely, debuginfo.

You need to replace `_LOCAL_TMP` with `((_LOCAL_1 as Variant ).FIELD: TY)` in debuginfo as well, making this a very specialized form of copy-prop, I guess?

cc @rust-lang/wg-mir-opt

Reply:
It would be nice to preserve debuginfo perfectly, but for that we might need a way to merge scopes or something, so that `SourceInfo` from all 3 statements is combined.

Effectively, the resulting `_LOCAL_0 = move _LOCAL_1` needs to be in a superposition of all the scopes.

Which might be hard due to the inlining.